xds/internal/balancer/clusterimpl/clusterimpl.go

```diff
@@ -360,22 +361,33 @@ func (scw *scWrapper) localityID() xdsinternal.LocalityID {
 func (b *clusterImplBalancer) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
 	clusterName := b.getClusterName()
 	newAddrs := make([]resolver.Address, len(addrs))
-	var lID xdsinternal.LocalityID
 	for i, addr := range addrs {
 		newAddrs[i] = xds.SetXDSHandshakeClusterName(addr, clusterName)
-		lID = xdsinternal.GetLocalityID(newAddrs[i])
 	}
 	var sc balancer.SubConn
+	scw := &scWrapper{}
 	oldListener := opts.StateListener
-	opts.StateListener = func(state balancer.SubConnState) { b.updateSubConnState(sc, state, oldListener) }
+	opts.StateListener = func(state balancer.SubConnState) {
+		b.updateSubConnState(sc, state, oldListener)
+		// Read connected address and call updateLocalityID() based on the connected
+		// address's locality. https://github.com/grpc/grpc-go/issues/7339
+		if gca, ok := internal.GetConnectedAddress.(func(balancer.SubConnState) (resolver.Address, bool)); ok {
+			if addr, ok := gca(state); ok {
+				lID := xdsinternal.GetLocalityID(addr)
+				if !lID.Empty() {
+					scw.updateLocalityID(lID)
+				} else if b.logger.V(2) {
+					b.logger.Infof("Locality ID for %v unexpectedly empty", addr)
```

Comment:
I would replace the `%v` here with a `%s`. The `resolver.Address` type implements a `String()` method which satisfies the `stringer` interface and therefore can used directly with the `%s` formatting directive.

Reply:
Done.